src/notifications/slices/subscriptions.ts

```diff
@@ -0,0 +1,208 @@
+  async (_, { dispatch }) => {
+    const { subscription, message } = await registerPushManager();
+    if (subscription) {
+      const deviceSubscription = transformDeviceSubscription(subscription);
+      dispatch(subscriptionUpdated({ subscription: deviceSubscription, message }));
+      const response = await createNotificationSubscription(deviceSubscription);
+      if (response.status === 'success') {
+        return response.data;
+      } else {
+        throw response.errors;
+      }
+    } else {
+      dispatch(subscriptionUpdated({ subscription: null, message }));
+      throw message;
+    }
+  }
+);
+
+export const unregisterDeviceFromPushNotfications = createAsyncThunk(
+  'notificationSubscriptions/unregisterDevice',
+  async (_, { getState }) => {
+    const state = getState() as State;
+    const { deviceSubscription } = state.notificationSubscriptions;
+    if (!deviceSubscription) {
+      return;
+    }
+    const subscription = subscriptionSelectors
+      .selectAll(state)
+      .find((sub) => sub.endpoint === deviceSubscription.endpoint);
+    if (!subscription) {
+      return;
+    }
+    const response = await destroyNotificationSubscription(subscription.id);
+    if (response.status !== 'success') {
+      throw response.errors;
+    }
+    const removed = await removeNotificationSubscription();
+    if (removed) {
+      subscriptionUpdated({ subscription: null, message: null });
+    }
+    return subscription.id;
+  }
+);
+
+export const toggleDevicePermission = createAsyncThunk('notificationSubscriptions/toggleDevicePermission', async () => {
+  const allowNotifications = await getNotificationPermission();
+  return allowNotifications;
+});
+
+export const resolveCurrentDeviceSubscription = createAsyncThunk(
+  'notificationSubscriptions/resolveCurrent',
+  async () => {
+    const subscription = await getNotificationSubscription();
+    if (subscription) {
+      const deviceSubscription = transformDeviceSubscription(subscription);
+      return deviceSubscription;
+    } else {
+      throw null;
+    }
+  }
+);
+
+interface IState {
+  loading: 'idle' | 'pending';
+  error: SerializedError | null;
+  entities: Record<number, ISubscription>;
+  deviceSubscription: IDeviceSubscription | null;
+  allowNotifications: boolean;
+  message: string | null;
+}
+
+const INITIAL_STATE: IState = {
+  loading: 'idle',
+  error: null,
+  entities: {},
+  deviceSubscription: null,
+  allowNotifications: resolveNotificationPermission(),
+  message: null,
+};
+
+const subscriptionsSlice = createSlice({
+  name: 'notificationSubscriptions',
+  initialState: subscriptionsAdapter.getInitialState(INITIAL_STATE),
+  reducers: {
+    subscriptionUpdated(
+      state,
+      action: PayloadAction<{ subscription: IDeviceSubscription | null; message: string | null }>
+    ) {
+      const { subscription, message } = action.payload;
+      state.message = message;
+      state.deviceSubscription = subscription;
+    },
```

Comment:
This is just a regular function defined in an object literal. Just as how you would do it in Vue components.

You can do the same with arrow functions, but it would be extra characters without any gain. I think it looks neat like that, and it is also how they do it in the Redux Toolkit docs :smile: